### PR TITLE
ENH: RangeIndex.sort_values now avoids materialization of values

### LINKED ISSUE #43584: ENH: `RangeIndex.sort_values` can avoid materialization of values.
#### Is your feature request related to a problem?

We can avoid the materialization of `RangeIndex` when `sort_values` is invoked on it. The reason being RangeIndex is always sorted.

#### Describe the solution you'd like

```python
>>> import pandas as pd
>>> pd.__version__
'1.3.3'
>>> idx = pd.RangeIndex(0, 10, 1)
>>> idx
RangeIndex(start=0, stop=10, step=1)
>>> idx.sort_values(ascending=True)
Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype='int64')  # Instead we should be returning RangeIndex(start=0, stop=10, step=1)

>>> idx.sort_values(ascending=False)
Int64Index([9, 8, 7, 6, 5, 4, 3, 2, 1, 0], dtype='int64')    #Instead we can return RangeIndex(start=10-1, start=0-1, step=-1), see below
>>> pd.RangeIndex(10-1, 0-1, -1)._values
array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])
```

#### API breaking implications

This would break the expected return types of `Index` of `DataFrame` & `Series` in some operations, but would result in speedup of operation as we would not be materializing the `RangeIndex`.

### PR DESCRIPTION
**Pull request by usersblock**
- [x] closes #43584
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry
What's new: On any instance of RangeIndex, sort_values will not sort as it is already sorted.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5206,8 +5206,29 @@ def sort_values(
         >>> idx.sort_values(ascending=False, return_indexer=True)
         (Int64Index([1000, 100, 10, 1], dtype='int64'), array([3, 1, 0, 2]))
         """
+        from pandas.core.indexes.range import RangeIndex
+
         idx = ensure_key_mapped(self, key)
 
+        if isinstance(self, RangeIndex):
```
Comment (jreback) [inline pandas/core/indexes/base.py:5213]: instead of this you can override `.sort_values` in RangeIndex to handle this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls add asv's that catch this case.

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -548,6 +548,34 @@ def equals(self, other: object) -> bool:
...
         return super().equals(other)
 
+    def sort_values(
+        self,
+        na_position: str = "last",
+        return_indexer: bool = False,
+        ascending: bool = True,
+        key: Callable | None = None,
+    ):
+        sorted_index = self
```
Comment (jreback) [inline pandas/core/indexes/range.py:558]: can you raise if key is not None here (and add a correesponding test).

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also some pre-commit issues with typing: https://github.com/pandas-dev/pandas/pull/43666/checks?check_run_id=3659974388

**Review by jreback — CHANGES_REQUESTED** [review]
also pls add a whatsnew note in perf improvements for 1.4

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (usersblock) [reply pandas/core/indexes/base.py]: There is a @final in the original function. Should I remove it?

**Comment by jreback** [comment]
@usersblock if you can address comments and merge master

**Comment by usersblock** [comment]
I still need to know whether I should remove the @final since it's blocking my override.

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/base.py]: yes

**Comment by jbrockmendel** [comment]
> I still need to know whether I should remove the @Final since it's blocking my override.

yes you should

**Review by jreback — CHANGES_REQUESTED** [review]
and pls add a whatsnew note in performance section

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (usersblock) [reply pandas/core/indexes/base.py]: Done

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
i think in test_algos.TestFactorize.test_factorize with this PR you could pass `exact=True` to `assert_index_equal`

**Comment by jreback** [comment]
thanks @usersblock very nice!

File: pandas/tests/indexes/ranges/test_range.py [hunk 3]
```diff
@@ -607,3 +607,11 @@ def test_isin_range(self, base):
         result = base.isin(values)
         expected = np.array([True, False])
         tm.assert_numpy_array_equal(result, expected)
+
+    def test_sort_values_key(self):
+        # GH#43666
+        sort_order = {8: 2, 6: 0, 4: 8, 2: 10, 0: 12}
+        values = RangeIndex(0, 10, 2)
+        result = values.sort_values(key=lambda x: x.map(sort_order))
+        expected = Index([4, 8, 6, 0, 2], dtype="int64")
```
Comment (jbrockmendel) [inline pandas/tests/indexes/ranges/test_range.py:616]: @usersblock where did you get this `expected` from?  it doesn't match what we'd get if we converted to Series before doing sort_values

**Pull request merged**